core/src/main/java/feign/RequestTemplate.java

```diff
@@ -535,6 +535,8 @@ public RequestTemplate headers(Map<String, Collection<String>> headers) {
    * @see Request#headers()
    */
   public Map<String, Collection<String>> headers() {
+    headers.forEach((key, value) -> value.removeIf(e -> e == null));
```

Comment:
I don't like the idea of modifying the existing map...

may be just filter and create  new one?!